src/passes/OptimizeInstructions.cpp

```diff
@@ -1835,17 +1835,46 @@ struct OptimizeInstructions
         Bits::getMaxBits(curr->ifFalse, this) <= 1) {
       // The condition and both arms are i32 booleans, which allows us to do
       // boolean optimizations.
-      Expression* x;
-      Expression* y;
 
       // x ? y : 0   ==>   x & y
-      if (matches(curr, select(any(&y), ival(0), any(&x)))) {
-        return builder.makeBinary(AndInt32, y, x);
+      if (matches(curr->ifFalse, ival(0))) {
+        return builder.makeBinary(AndInt32, curr->ifTrue, curr->condition);
       }
 
       // x ? 1 : y   ==>   x | y
-      if (matches(curr, select(ival(1), any(&y), any(&x)))) {
-        return builder.makeBinary(OrInt32, y, x);
+      if (matches(curr->ifTrue, ival(1))) {
+        return builder.makeBinary(OrInt32, curr->ifFalse, curr->condition);
+      }
+
+      // Some operations require us to add an eqz, like
+      // x ? 0 : y   ==>   !x & y
+      // We replace the select with an and, which does not change code size, and
+      // then reduce code size by removing the number (2 bytes) and adding an
+      // eqz (one byte). However, the eqz adds work, so it is not obvious that
+      // this is beneficial to do - only do it when we know we can fold away the
+      // eqz.
+      // TODO: Perhaps also do this even if we can't remove the eqz, if we are
+      //       optimizing for size? We should measure the performance aspect
+      //       first though, to see if we need that.
```

Comment:
I guess instead `eqz` it could be `x ^ 1` which may be a faster but bigger in term of code size

Reply:
Hmm, I'd expect the VM to optimize that, if it were faster. Otherwise every eqz should be replaced with that larger pattern.

Reply:
replacing `eqz(x)` to `x ^ 1` possible only if `getMaxBits(x) == 1` so it's not always possible and I guess not always clear for VMs

Reply:
Ah, good point. Might be worth looking at then.

Reply:
However now I'm not sure. We already have reversed rule in OptimizeInstructions for: `x ^ 1` -> `eqz(x)`, if `getMaxBits(x) == 1`. But I guess we could hide it under opt options. I have an assumption that eqz(x) could be best option if it combined with if, br_if or select but suboptimal for rest cases